internal/manager/controllerdef.go

```diff
@@ -331,124 +323,123 @@ func setupControllers(
+						Scheme:               mgr.GetScheme(),
+						DataplaneClient:      dataplaneClient,
+						PublishServiceRef:    c.PublishService.OrEmpty(),
+						PublishServiceUDPRef: c.PublishServiceUDP,
+						WatchNamespaces:      c.WatchNamespaces,
+						CacheSyncTimeout:     c.CacheSyncTimeout,
+						ReferenceIndexers:    referenceIndexers,
+					},
+					&gateway.HTTPRouteReconciler{
+						Client:           mgr.GetClient(),
+						Log:              ctrl.Log.WithName("controllers").WithName("HTTPRoute"),
+						Scheme:           mgr.GetScheme(),
+						DataplaneClient:  dataplaneClient,
+						CacheSyncTimeout: c.CacheSyncTimeout,
+					},
 				},
-				restMapper,
-			),
-			Controller: &gateway.HTTPRouteReconciler{
-				Client:               mgr.GetClient(),
-				Log:                  ctrl.Log.WithName("controllers").WithName("HTTPRoute"),
-				Scheme:               mgr.GetScheme(),
-				DataplaneClient:      dataplaneClient,
-				EnableReferenceGrant: referenceGrantsEnabled,
-				CacheSyncTimeout:     c.CacheSyncTimeout,
 			},
 		},
 		// ---------------------------------------------------------------------------
 		// Gateway API Controllers - Alpha APIs
 		// ---------------------------------------------------------------------------
 		{
-			Enabled: referenceGrantsEnabled,
-			Controller: &gateway.ReferenceGrantReconciler{
-				Client:           mgr.GetClient(),
-				Log:              ctrl.Log.WithName("controllers").WithName("ReferenceGrant"),
-				Scheme:           mgr.GetScheme(),
-				DataplaneClient:  dataplaneClient,
+			Enabled: featureGates[featuregates.GatewayAlphaFeature],
+			Controller: &crds.DynamicCRDController{
+				Manager:          mgr,
+				Log:              ctrl.Log.WithName("controllers").WithName("Dynamic/gatewayv1alpha2"),
 				CacheSyncTimeout: c.CacheSyncTimeout,
-			},
-		},
-		{
-			Enabled: featureGates[featuregates.GatewayAlphaFeature] && ShouldEnableCRDController(
-				schema.GroupVersionResource{
-					Group:    gatewayv1alpha2.GroupVersion.Group,
-					Version:  gatewayv1alpha2.GroupVersion.Version,
-					Resource: "udproutes",
+				RequiredCRDs: []schema.GroupVersionResource{
+					{
+						Group:    gatewayv1beta1.GroupVersion.Group,
+						Version:  gatewayv1beta1.GroupVersion.Version,
+						Resource: "referencegrants",
+					},
+					{
+						Group:    gatewayv1alpha2.GroupVersion.Group,
+						Version:  gatewayv1alpha2.GroupVersion.Version,
+						Resource: "udproutes",
+					},
+					{
+						Group:    gatewayv1alpha2.GroupVersion.Group,
+						Version:  gatewayv1alpha2.GroupVersion.Version,
+						Resource: "tcproutes",
+					},
+					{
+						Group:    gatewayv1alpha2.GroupVersion.Group,
+						Version:  gatewayv1alpha2.GroupVersion.Version,
+						Resource: "tlsroutes",
+					},
+					{
+						Group:    gatewayv1alpha2.GroupVersion.Group,
+						Version:  gatewayv1alpha2.GroupVersion.Version,
+						Resource: "grpcroutes",
+					},
```

Comment:
I wonder if grouping all of those is a good path forward 🤔 Do we actually need to group them? Does it make sense for the end user to need to add `GRPCRoute` CRD to have the functionality of `TCPRoute`?

Reply:
Conceptually, grouping them on the API level: `gatewayv1alpha2` or `gatewayv1beta` makes sense, but from the use case I described above it might not 🤔

Reply:
Yeah, you're right, that's a good point. I made it granular so that every controller is handled separately, PTAL.